sendermode.go

```diff
@@ -183,44 +187,57 @@ func (fm *Frontman) sendResultsChanToHubQueue(resultsChan *chan Result) error {
 						}
 						logrus.Errorf("postResultsToHub error: %s", err.Error())
 					}
-
-					var m runtime.MemStats
-					runtime.ReadMemStats(&m)
-					logrus.Debugf("Sender queue: Alloc = %v, TotalAlloc = %v, Sys = %v, NumGC = %v", m.Alloc/1024, m.TotalAlloc/1024, m.Sys/1024, m.NumGC)
 				}(sendResults)
 			} else {
-				logrus.Debugf("Sender queue: nothing to do. outgoing queue empty.")
+				logrus.Infof("sendResultsChanToHubQueue: nothing to do. outgoing queue empty.")
 			}
 		}
 
 		if time.Since(lastQueueStatsWrite) >= writeQueueStatsInterval {
 			lastQueueStatsWrite = time.Now()
+			fm.writeQueueStats(len(results))
+		}
 
-			data, err := json.Marshal(map[string]int{
-				"checks_queue":       len(fm.checks),
-				"checks_in_progress": fm.ipc.len(),
-				"results_queue":      len(results)})
-			if err != nil {
+		select {
+		case <-interrupt:
+			logrus.Infof("sendResultsChanToHubQueue interrupt caught, should return. results %v", len(results))
+			if err := fm.postResultsToHub(results); err != nil {
 				logrus.Error(err)
-			} else if fm.Config.QueueStatsFile != "" {
-				go func(b []byte) {
-					f, err := os.Create(fm.Config.QueueStatsFile)
-					if err != nil {
-						logrus.Error(err)
-						return
-					}
-					defer f.Close()
-					_, err = f.Write(b)
-					if err != nil {
-						logrus.Error(err)
-					}
-				}(data)
 			}
+			fm.writeQueueStats(0)
+			return
+		default:
 		}
+	}
+}
 
-		if shouldReturn && len(results) == 0 {
-			fm.TerminateQueue.Done()
-			return nil
-		}
+func (fm *Frontman) writeQueueStats(resultsLen int) {
+	if fm.Config.QueueStatsFile == "" {
+		return
+	}
+
+	fm.checksLock.Lock()
+	data, err := json.Marshal(map[string]int{
+		"checks_queue":       len(fm.checks),
+		"checks_in_progress": fm.ipc.len(),
+		"results_queue":      resultsLen})
+	fm.checksLock.Unlock()
+
+	if err != nil {
+		logrus.Error(err)
+		return
 	}
+
+	go func(b []byte) {
+		f, err := os.Create(fm.Config.QueueStatsFile)
+		if err != nil {
+			logrus.Error(err)
```

Comment:
but if this go routine stops because of error, will main go routine be aware of it? will it wait forever or expect everything is going well? I would expect the whole program to exit with exit code !=0, if one go routine failed because of some critical problem (e.g. fs is not writeable). There should be some sync mechanism between main and sub goroutines, e.g. channels or context.

Reply:
Main goroutine doesnt care if `writeQueueStats` fails. We simply wont have more stats written to disk but a log entry should tell us why.
We do use channels for the goroutine communication that matters.

Reply:
sure